pkg/readers/bufferedfilereader.go

```diff
@@ -0,0 +1,91 @@
+package readers
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/trufflesecurity/trufflehog/v3/pkg/context"
+	bufferedfilewriter "github.com/trufflesecurity/trufflehog/v3/pkg/writers/buffered_file_writer"
+)
+
+// Compile time check to ensure that bufferedFileReader implements io.ReaderAt.
+var _ io.ReaderAt = (*bufferedFileReader)(nil)
+
+// bufferedFileReader provides random access read, seek, and close capabilities on top of the BufferedFileWriter.
+// It combines the functionality of BufferedFileWriter for buffered writing, io.ReadSeekCloser for
+// random access reading and seeking.
+type bufferedFileReader struct {
+	bufWriter *bufferedfilewriter.BufferedFileWriter
+	reader    io.ReadSeekCloser
+}
+
+// NewBufferedFileReader initializes a bufferedFileReader from an io.Reader by using
+// the BufferedFileWriter's functionality to read and store data, then setting up a io.ReadSeekCloser
+// for random access reading and seeking.
+// Close should be called when the bufferedFileReader is no longer needed.
+// It returns the initialized bufferedFileReader and any error encountered during the process.
+func NewBufferedFileReader(ctx context.Context, r io.Reader) (*bufferedFileReader, error) {
+	writer, err := bufferedfilewriter.NewFromReader(ctx, r)
+	if err != nil {
+		return nil, fmt.Errorf("error creating bufferedFileReader: %w", err)
+	}
+
+	// Ensure that the BufferedFileWriter is in read-only mode.
+	if err := writer.CloseForWriting(); err != nil {
+		return nil, err
+	}
+
+	reader, err := writer.ReadCloser()
```

Comment:
This is using the `BufferedFileWriter` [ReadCloser](https://github.com/trufflesecurity/trufflehog/blob/a5f04e65f78191ce388f93c9cb2cec29feea7421/pkg/writers/buffer_writer/bufferwriter.go#L72-L81) implementation, correct? Doesn't that load the entire thing into memory, contradicting the first point? I feel like I'm missing something.

> 1. This will provide a way to create a custom reader that can accept an `io.Reader` with any amount of data without worrying about holding all the data in memory.

Reply:
It should not. It would only do that _if_ the writer is under the threshold for requiring a file. If it is not, we return an _autoDeletingFile_.

Reply:
Oops, I was looking at the `BufferedWriter` implementation, not the [BufferedFileWriter](https://github.com/trufflesecurity/trufflehog/blob/a5f04e65f78191ce388f93c9cb2cec29feea7421/pkg/writers/buffered_file_writer/bufferedfilewriter.go#L201) implementation.

Reply:
Yea... 😢 I could have done a better job distinguishing between the two. I'm open to renaming them to avoid confusion moving forward. Do you have any suggestions?

Reply:
I think the names are fine, I just made a mistake 🙂